tsbridge/metric.go

```diff
@@ -55,22 +56,41 @@ type StackdriverAdapter interface {
 }
 
 // UpdateAllMetrics updates all metrics listed in a given config.
-func UpdateAllMetrics(ctx context.Context, c *Config, sd StackdriverAdapter, s *StatsCollector) (errors []string) {
+func UpdateAllMetrics(ctx context.Context, c *Config, sd StackdriverAdapter, parallelism int, s *StatsCollector) (errors []string) {
 	oldestWrite := time.Now()
 	defer func(start time.Time) {
 		stats.Record(ctx, s.TotalImportLatency.M(int64(time.Since(start)/time.Millisecond)))
 		stats.Record(ctx, s.OldestMetricAge.M(int64(time.Since(oldestWrite)/time.Millisecond)))
 	}(time.Now())
 
+	errchan := make(chan string, len(c.Metrics()))
+	sem := make(chan bool, parallelism)
+	var wg sync.WaitGroup
+
+	for _, m := range c.Metrics() {
+		sem <- true
+		wg.Add(1)
+		go func(metric *Metric) {
+			defer wg.Done()
+			defer func() { <-sem }()
+
+			err := metric.Update(ctx, sd, s)
+			if err != nil {
+				errchan <- err.Error()
+			}
+		}(m)
+	}
+	wg.Wait()
+	close(errchan)
+
 	for _, m := range c.Metrics() {
-		err := m.Update(ctx, sd, s)
-		if err != nil {
-			errors = append(errors, err.Error())
-		}
 		if m.Record.LastUpdate.Before(oldestWrite) {
```

Comment:
Why isn't this done in the above loop?

Reply:
Because `m.Record.LastUpdate` gets updated by one of those goroutines that get started behind a `SyncGroup`, so we need to wait for all of them to finish first. I've added a comment clarifying this.